flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java

```diff
@@ -178,36 +164,96 @@ private static ReporterSetup createReporterSetup(String reporterName, MetricConf
+
+		// scan entire configuration for keys starting with METRICS_REPORTER_PREFIX and determine the set of enabled reporters
+		for (String key : configuration.keySet()) {
+			if (key.startsWith(ConfigConstants.METRICS_REPORTER_PREFIX)) {
+				Matcher matcher = reporterClassPattern.matcher(key);
+				if (matcher.matches()) {
+					String reporterName = matcher.group(1);
+					if (includedReporters.isEmpty() || includedReporters.contains(reporterName)) {
+						if (namedOrderedReporters.contains(reporterName)) {
+							LOG.warn("Duplicate class configuration detected for reporter {}.", reporterName);
+						} else {
+							namedOrderedReporters.add(reporterName);
+						}
+					} else {
+						LOG.info("Excluding reporter {}, not configured in reporter list ({}).", reporterName, includedReportersString);
+					}
+				}
+			}
+		}
+		return namedOrderedReporters;
+	}
+
+	private static Map<String, MetricReporterFactory> loadAvailableReporterFactories(PluginManager pluginManager) {
 		final Map<String, MetricReporterFactory> reporterFactories = new HashMap<>(2);
-		final Iterator<MetricReporterFactory> factoryIterator = serviceLoader.iterator();
+		final Iterator<MetricReporterFactory> factoryIterator = getAllReporterFactories(pluginManager);
+		LOG.debug("All available factories (from both SPIs and Plugins):");
+		getAllReporterFactories(pluginManager).forEachRemaining(i -> LOG.debug(i.toString()));
 		// do not use streams or for-each loops here because they do not allow catching individual ServiceConfigurationErrors
 		// such an error might be caused if the META-INF/services contains an entry to a non-existing factory class
 		while (factoryIterator.hasNext()) {
 			try {
 				MetricReporterFactory factory = factoryIterator.next();
-				reporterFactories.put(factory.getClass().getName(), factory);
+				String factoryClassName = factory.getClass().getName();
+				MetricReporterFactory existingFactory = reporterFactories.get(factoryClassName);
+				if (existingFactory == null){
+					reporterFactories.put(factoryClassName, factory);
+					LOG.warn(new File(factory.getClass().getProtectionDomain().getCodeSource().getLocation()
+						.toURI()).getCanonicalPath());
+				} else {
+					//TODO: use path information below, when Plugin Classloader stops always prioritizing factories from /lib
+//					String jarPath1 = new File(existingFactory.getClass().getProtectionDomain().getCodeSource().getLocation()
+//						.toURI()).getCanonicalPath();
+//					String jarPath2 = new File(factory.getClass().getProtectionDomain().getCodeSource().getLocation()
+//						.toURI()).getCanonicalPath();
+//					LOG.warn("Multiple implementations of the same reporter were found: \n {} and \n{}", jarPath1, jarPath2);
+					LOG.warn("Multiple implementations of the same reporter were found in 'lib' and 'plugins' directories for {}. It is recommended to remove redundant reporter JARs to resolve used versions' ambiguity.", factoryClassName);
```

Comment:
Is it guaranteed that they are in opt _and_ lib? What if 2 plugins specified the same factory class?

Reply:
But this would technically be "multiple implementations of the same reporter"? The message does not tell explicitly that one of them is in lib and another in plugins - just that while searching those two directories, multiple implementations were found.

Reply:
Strictly speaking it then should be 'and/or', because 'or' could mean that there are other two jars both either in 'lib' or 'plugins'. I think and was more fitting, but I am also ok with and/or